tf_quant_finance/black_scholes/approximations/american_option.py

```diff
@@ -0,0 +1,424 @@
+          volatilities, strikes, expiries, risk_free_rates, cost_of_carries,
+          spots=spots, forwards=forwards, is_call_options=is_call_options,
+          dtype=dtype, name=name)))],
+    default=lambda: _split_out(
+          volatilities, strikes, expiries, spots=spots, forwards=forwards,
+          risk_free_rates=risk_free_rates, cost_of_carries=cost_of_carries,
+          is_call_options=is_call_options, dtype=dtype, name=name),
+    exclusive=True)
+
+
+def _split_out(volatilities,
+               strikes,
+               expiries,
+               risk_free_rates,
+               cost_of_carries,
+               spots=None,
+               forwards=None,
+               is_call_options=None,
+               dtype=None,
+               name="american_price"):
+  """
+  Splits prices out to European and American batches, then combines the result.
+  """
+
+  if forwards is not None:
+    t = tf.broadcast_to(tf.convert_to_tensor(
+      expiries, dtype=dtype), forwards.shape, name="expiries")
+    s = tf.convert_to_tensor(
+      forwards * tf.exp(- cost_of_carries * t),
+      dtype=dtype,
+      name="spots")
+  else:
+    t = tf.broadcast_to(tf.convert_to_tensor(
+      expiries, dtype=dtype), spots.shape, name="expiries")
+    s = tf.convert_to_tensor(spots, dtype=dtype, name="spots")
+
+  b_less_r = risk_free_rates > cost_of_carries
+  b_more_eq_r = risk_free_rates <= cost_of_carries
+  am_indices = tf.where(b_less_r)
+  eu_indices = tf.where(b_more_eq_r)
+  shape = tf.shape(s, out_type=am_indices.dtype)
+
+  strikes = tf.broadcast_to(tf.convert_to_tensor(
+    strikes, dtype=dtype), shape, name="strikes")
+  volatilities = tf.broadcast_to(tf.convert_to_tensor(
+    volatilities, dtype=dtype), shape, name="volatilities")
+  cost_of_carries = tf.broadcast_to(tf.convert_to_tensor(
+    cost_of_carries, dtype=dtype), shape, name="cost_of_carries")
+  risk_free_rates = tf.broadcast_to(tf.convert_to_tensor(
+    risk_free_rates, dtype=dtype), shape, name="risk_free_rates")
+
+  if is_call_options is not None:
+    is_call_options = tf.broadcast_to(tf.convert_to_tensor(
+      is_call_options), s.shape, name="risk_free_rates")
+    is_call_eu = tf.gather_nd(is_call_options, eu_indices)
+    is_call_am = tf.gather_nd(is_call_options, am_indices)
+  else:
+    is_call_eu, is_call_am = None, None
+
+  x_eu = tf.gather_nd(strikes, eu_indices)
+  sigma_eu = tf.gather_nd(volatilities, eu_indices)
+  s_eu = tf.gather_nd(s, eu_indices)
+  t_eu = tf.gather_nd(t, eu_indices)
+  r_eu = tf.gather_nd(risk_free_rates, eu_indices)
+  b_eu = tf.gather_nd(cost_of_carries, eu_indices)
+
+  x_am = tf.gather_nd(strikes, am_indices)
+  sigma_am = tf.gather_nd(volatilities, am_indices)
+  s_am = tf.gather_nd(s, am_indices)
+  t_am = tf.gather_nd(t, am_indices)
+  r_am = tf.gather_nd(risk_free_rates, am_indices)
+  b_am = tf.gather_nd(cost_of_carries, am_indices)
+
+  eu_prices = tff.black_scholes.option_price(
```

Comment:
You should not be able to access the library before it was compiled. Please import the relevant module directly, i.e.,
from tf_quant_finance.black_scholes import vanilla_prices